src/main/java/org/datadog/jmxfetch/util/JmxfetchRmiClientSocketFactory.java

```diff
@@ -0,0 +1,30 @@
+package org.datadog.jmxfetch.util;
+
+import java.io.IOException;
+import java.net.Socket;
+import java.rmi.server.RMIClientSocketFactory;
+import java.rmi.server.RMISocketFactory;
+import javax.rmi.ssl.SslRMIClientSocketFactory;
+
+public class JmxfetchRmiClientSocketFactory implements RMIClientSocketFactory {
+    private final int timeoutMs;
+    private RMIClientSocketFactory factory;
+
+    /** 
+     * JmxfetchRmiClientSocketFactory constructor with socket timeout (milliseconds) and 
+     * a flag to enable/disable SSL.
+     */
+    public JmxfetchRmiClientSocketFactory(int timeoutMs, boolean ssl) {
+        this.timeoutMs = timeoutMs;
+        this.factory = ssl ? new SslRMIClientSocketFactory()
+                : RMISocketFactory.getDefaultSocketFactory();
+    }
+
+    @Override
+    public Socket createSocket(String host, int port) throws IOException {
+        Socket socket = factory.createSocket(host, port);
+        socket.setSoTimeout(timeoutMs);
```

Comment:
In the past we had two different timeout values:
- `sun.rmi.transport.proxy.connectTimeout` (connect TO)
- `sun.rmi.transport.tcp.responseTimeout` (read TO)

This socket-level, and more explicit `SO_TIMEOUT` is probably fine for all cases, but we may be potentially changing behavior on new connections by dropping the `rmi_connection_timeout`. It does feel like a bit of a redundant option, but I'm wondering if we didn't have to set that for some weird edge-case.

Also, I may be missing something!

Reply:
I removed the usage of `sun.rmi.transport.tcp.responseTimeout` because it sets `SO_TIMEOUT` later in the whole thing lifecyle (as per [1](http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/jdk8u232-ga/src/share/classes/sun/rmi/transport/tcp/TCPChannel.java#l99) and [2](http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/jdk8u232-ga/src/share/classes/sun/rmi/transport/tcp/TCPChannel.java#l291)) and some earlier operation would occur without having the `SO_TIMEOUT` set.

AFAIK the latter option `sun.rmi.transport.proxy.connectTimeout` _should_ not have impact on our base case as far as my test went, we exit the socket factory [here](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/master/jdk/src/share/classes/sun/rmi/transport/proxy/RMIMasterSocketFactory.java#L148), before using the property value. I had a quick look in the code base history I did not find where we were using this property...

Reply:
@truthbk I did something similar to the mailbox code (:trollface:) thus we are getting back the `rmi_connection_timeout`, however it's fully located in the socket factory so any subsequent connection not going through `RemoteConnection`  will still be subject to the timeout and I think all case are covered (i.e. SSL/non-SSL, initial connection, subsequent connection directly by the RMI subsystem, timeout during socket lifetime).
@sirianni it should bring an effective & generic solution to the issue #279.